sdk/lib/rtl/rangelist.c

```diff
@@ -13,546 +13,1556 @@
-        if (NewEntry == NULL)
-            return STATUS_INSUFFICIENT_RESOURCES;
+        MergedRtlEntry = CONTAINING_RECORD(Iterator->MergedHead, RTLP_RANGE_LIST_ENTRY, Merged.ListHead);
+
+        if (MoveForwards)
+        {
+            ListEntry = MergedRtlEntry->ListEntry.Flink;
+        }
+        else
+        {
+            ListEntry = MergedRtlEntry->ListEntry.Blink;
+        }
 
-        RtlCopyMemory(&NewEntry->Range,
-                      &Current->Range,
-                      sizeof(RTL_RANGE));
+        NextRtlEntry = RtlpEntryFromLink(ListEntry);
+        DPRINT("RtlGetNextRange: %p, %p, %p [%I64X-%I64X]\n", MergedRtlEntry, ListEntry, NextRtlEntry, NextRtlEntry->Start, NextRtlEntry->End);
 
-        InsertTailList(&CopyRangeList->ListHead,
-                       &NewEntry->Entry);
+        Iterator->MergedHead = NULL;
+    }
+
+    if (&NextRtlEntry->ListEntry == Iterator->RangeListHead)
+    {
+        DPRINT("RtlGetNextRange: return STATUS_NO_MORE_ENTRIES\n");
+        Iterator->Current = NULL;
+        *OutRange = NULL;
+        return STATUS_NO_MORE_ENTRIES;
+    }
 
-        CopyRangeList->Count++;
+    if (NextRtlEntry->PrivateFlags & RTLP_ENTRY_IS_MERGED)
+    {
+        ASSERT(!Iterator->MergedHead);
+        Iterator->MergedHead = &NextRtlEntry->Merged.ListHead;
 
-        Entry = Entry->Flink;
+        if (MoveForwards)
+        {
+            ListEntry = NextRtlEntry->Merged.ListHead.Flink;
+        }
+        else
+        {
+            ListEntry = NextRtlEntry->Merged.ListHead.Blink;
+        }
+    }
+    else
+    {
+        ListEntry = &NextRtlEntry->ListEntry;
     }
 
-    CopyRangeList->Stamp++;
+    NextRtlEntry = RtlpEntryFromLink(ListEntry);
+    DPRINT("RtlGetNextRange: %p, %p, *OutRange %X [%I64X-%I64X]\n", Iterator->Current, Iterator->MergedHead, NextRtlEntry, NextRtlEntry->Start, NextRtlEntry->End);
+
+    Iterator->Current = NextRtlEntry;
+    *OutRange = (PRTL_RANGE)NextRtlEntry;
 
     return STATUS_SUCCESS;
 }
 
-
-/**********************************************************************
- * NAME							EXPORTED
- * 	RtlDeleteOwnersRanges
+/**
+ * @name RtlGetLastRange
  *
- * DESCRIPTION
- *	Delete all ranges that belong to the given owner.
+ * Retrieves the last range of a range list.
  *
- * ARGUMENTS
- *	RangeList	Pointer to the range list.
- *	Owner		User supplied value that identifies the owner
- *			of the ranges to be deleted.
+ * @param RangeList
+ * Pointer to the range list.
  *
- * RETURN VALUE
- *	Status
+ * @param Iterator
+ * Pointer to a user supplied list state buffer.
  *
- * @implemented
+ * @param OutRange
+ * Address of pointer to the last range.
+ *
+ * @return
+ * Returns NTSTATUS.
  */
+CODE_SEG("PAGE")
+NTSYSAPI
 NTSTATUS
 NTAPI
-RtlDeleteOwnersRanges(IN OUT PRTL_RANGE_LIST RangeList,
-                      IN PVOID Owner)
+RtlGetLastRange(
+    _In_ PRTL_RANGE_LIST RangeList,
+    _Inout_ PRTL_RANGE_LIST_ITERATOR Iterator,
+    _Out_ PRTL_RANGE * OutRange)
```

Comment:
Fixing the annotation:
```suggestion
    _Outptr_ PRTL_RANGE *Range)
```
and similarly to the other function, for consistency, renaming the parameter to just `Range`.